lib/Sema/CSSimplify.cpp

```diff
@@ -1609,7 +1616,7 @@ static bool matchFunctionRepresentations(FunctionTypeRepresentation rep1,
   case ConstraintKind::OneWayEqual:
   case ConstraintKind::OneWayBindParam:
   case ConstraintKind::DefaultClosureType:
-    return false;
+    return (rep1 == rep2) && clangTypeMismatch;
```

Comment:
I think this should still return `false` since there is no way functions could be found in this position.

Reply:
I wasn't entirely sure about this one, because I saw that there is an `llvm_unreachable` in `matchFunctionTypes` after this. Does it make sense for this to be an `llvm_unreachable` instead?

Reply:
Since the switch is exhaustive I think we can remove that `llvm_unreachable`, adding new ConstraintKind would cause an error here.

Reply:
Changed back to `return false;`.

Reply:
Removing this seems to have caused a bunch of test failures for assignments, it is too permissive now.

```
let _ : @convention(c) (Int) -> Int = f1!
/Users/varun/foss-swift/swift/test/Sema/clang_fn_type_mismatch.swift:12:4: error: expected error not produced
// expected-error@-1{{cannot convert value of type '@convention(c, cType: "size_t (*)(size_t)") (Int) -> Int' to specified type '@convention(c) (Int) -> Int'}}
~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
In all those cases, this function ends up getting a `ConstraintKind::Conversion`. I do think for that kind of conversion, change Clang types should be disallowed.

Reply:
So if I modify this to:

```
  case ConstraintKind::Conversion:
    return ((rep1 == rep2) && clangTypeMismatch);
```

That fixes the issue. I don't know if it is the right fix though.

Reply:
Hm, I think Subtype, Conversion, ArgumentConversion and OperatorArgumentConversion should all be handled the same way